libs/vec/src/main21/java/org/elasticsearch/vec/internal/Int7SQVectorScorerSupplier.java

```diff
@@ -0,0 +1,219 @@
+import static org.apache.lucene.index.VectorSimilarityFunction.MAXIMUM_INNER_PRODUCT;
+import static org.apache.lucene.util.quantization.ScalarQuantizedVectorSimilarity.fromVectorSimilarity;
+
+public abstract sealed class Int7SQVectorScorerSupplier implements RandomVectorScorerSupplier {
+
+    final int dims;
+    final int maxOrd;
+    final float scoreCorrectionConstant;
+    final IndexInput input;
+    final RandomAccessQuantizedByteVectorValues values; // to support ordToDoc/getAcceptOrds
+    final ScalarQuantizedVectorSimilarity fallbackScorer;
+
+    final MemorySegment segment;
+    final MemorySegment[] segments;
+    final long offset;
+    final int chunkSizePower;
+    final long chunkSizeMask;
+
+    protected Int7SQVectorScorerSupplier(
+        IndexInput input,
+        RandomAccessQuantizedByteVectorValues values,
+        float scoreCorrectionConstant,
+        ScalarQuantizedVectorSimilarity fallbackScorer
+    ) {
+        this.input = input;
+        this.values = values;
+        this.dims = values.dimension();
+        this.maxOrd = values.size();
+        this.scoreCorrectionConstant = scoreCorrectionConstant;
+        this.fallbackScorer = fallbackScorer;
+
+        this.segments = IndexInputUtils.segmentArray(input);
+        if (segments.length == 1) {
+            segment = segments[0];
+            offset = 0L;
+        } else {
+            segment = null;
+            offset = IndexInputUtils.offset(input);
+        }
+        this.chunkSizePower = IndexInputUtils.chunkSizePower(input);
+        this.chunkSizeMask = IndexInputUtils.chunkSizeMask(input);
+    }
+
+    protected final void checkOrdinal(int ord) {
+        if (ord < 0 || ord > maxOrd) {
+            throw new IllegalArgumentException("illegal ordinal: " + ord);
+        }
+    }
+
+    final float scoreFromOrds(int firstOrd, int secondOrd) throws IOException {
+        checkOrdinal(firstOrd);
+        checkOrdinal(secondOrd);
+
+        final int length = dims;
+        long firstByteOffset = (long) firstOrd * (length + Float.BYTES);
+        long secondByteOffset = (long) secondOrd * (length + Float.BYTES);
+
+        MemorySegment firstSeg = segmentSlice(firstByteOffset, length);
+        input.seek(firstByteOffset + length);
+        float firstOffset = Float.intBitsToFloat(input.readInt());
+
+        MemorySegment secondSeg = segmentSlice(secondByteOffset, length);
+        input.seek(secondByteOffset + length);
+        float secondOffset = Float.intBitsToFloat(input.readInt());
+
+        if (firstSeg != null && secondSeg != null) {
+            return scoreFromSegments(firstSeg, firstOffset, secondSeg, secondOffset);
+        } else {
+            return fallbackScore(firstByteOffset, secondByteOffset);
+        }
+    }
+
+    abstract float scoreFromSegments(MemorySegment a, float aOffset, MemorySegment b, float bOffset);
+
+    protected final float fallbackScore(long firstByteOffset, long secondByteOffset) throws IOException {
```

Comment:
yeah, this is not used all that much but helps keep things reasonably straightforward on JDK 21. For 22 we'll get rid of it when adding support for search time - since we'll be able to pin on-heap vectors.